python/rmm/_lib/device_buffer.pyx

```diff
@@ -92,39 +95,40 @@ cdef class DeviceBuffer:
         return buf
 
     @staticmethod
-    @cython.boundscheck(False)
-    cdef DeviceBuffer c_frombytes(const unsigned char[::1] b,
+    cdef DeviceBuffer c_to_device(const unsigned char[::1] b,
                                   uintptr_t stream=0):
-        if b is None:
-            raise TypeError(
-                "Argument 'b' has incorrect type"
-                " (expected bytes, got NoneType)"
+        return to_device(b, stream)
+
+    @staticmethod
+    def to_device(const unsigned char[::1] b, uintptr_t stream=0):
+        return to_device(b, stream)
+
+    cpdef copy_to_host(self, unsigned char[::1] hb=None, uintptr_t stream=0):
+        cdef const device_buffer* dbp = self.c_obj.get()
+        cdef size_t s = dbp.size()
+
+        if hb is None:
+            # NumPy leverages huge pages under-the-hood,
+            # which speeds up the copy from device to host.
+            hb = np.empty((s,), dtype="u1")
+        elif len(hb) < s:
+            raise ValueError(
+                "Argument `hb` is too small. Need space for %i bytes." % s
             )
 
-        cdef uintptr_t p = <uintptr_t>&b[0]
-        cdef size_t s = len(b)
-        return DeviceBuffer(ptr=p, size=s, stream=stream)
+        with nogil:
+            copy_ptr_to_host(<uintptr_t>dbp.data(), hb[:s], stream)
 
-    @staticmethod
-    def frombytes(const unsigned char[::1] b, uintptr_t stream=0):
-        return DeviceBuffer.c_frombytes(b, stream)
+        return hb
```

Comment:
In the situation where someone passes in a host buffer to copy into this feels weird where it returns a host buffer object where it's up to them to determine is a reference to the host buffer object they pass in. Could we make this return `None` if they pass in a host buffer, or return a newly allocated host buffer if they don't pass one in?

Reply:
Good question. Also thought about this initially and found Numba returned in all cases. Not that that is a good argument for keeping it. That said, I could see how this would be handy...

```python
In [1]: import rmm

In [2]: import pandas

In [3]: rmm.reinitialize(pool_allocator=True, 
   ...:                  initial_pool_size=int(2 * 2**30))
Out[3]: 0

In [4]: rdb = rmm.DeviceBuffer(size=32)

In [5]: s = pandas.Series(rdb.copy_to_host(bytearray(rdb.nbytes)))
```

Reply:
That feels incredibly awkward from my perspective and doesn't fit along the lines of typical `numpy` / `Pandas` behavior of when inplace, return `None`.

Reply:
I don't think that is true.

```python
In [1]: import numpy as np                                                      

In [2]: a = np.ones((2, 3))                                                     

In [3]: np.multiply(a, 2, out=a)                                                
Out[3]: 
array([[2., 2., 2.],
       [2., 2., 2.]])
```